_posts/2021-01-05-overlappable-instances.markdown

```diff
@@ -0,0 +1,248 @@
+---
+layout: post
+title:  "Overlappable instances"
+date:   2021-01-05 00:00:12 +0100
+categories: haskell effects
+---
+There are many patterns, styles and libraries which can be chosen for dealing with effects in Haskell. Today I'll try to bring closer one of them.
+
+Let's start off with a very simple program written using tagless final style.
+The program inserts a value "1" into the store under the key "key1" and then tries to retrieve it using operations defined in the `Store` type class. It logs what is happening while doing the above thanks to the `Logger`.
+The instances needed for `IO` are added only to make the whole thing compile, the implementations aren't very useful (for now!)
+
+
+```haskell
+{-# LANGUAGE DerivingStrategies,UndecidableInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving, FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses, FlexibleInstances #-}
+module Effects.Overlapping where
+
+import Prelude hiding (log)
+
+class Store m where
+  put :: k -> a -> m ()
+  get :: k -> m (Maybe a)
```

Comment:
In practice, `k` and `a` would likely be arguments to the class, so you'd have `Store m k a`. Otherwise, the `k` and `a` in `get :: k -> m -> (Maybe a)` could be different from the `k` and `a` in `put :: k -> a -> m ()`.
However, `Store m k a` is a bit of a mouthful, so I'd replace `k` with a concrete type.
```suggestion
type Key = String

class Store m a where
  put :: Key -> a -> m ()
  get :: Key -> m (Maybe a)
```

Reply:
alternatively, if you go with `FileSystem`, you can make both the key and the value concrete:
```suggestion
class FileSystem m where
  writeFile :: FilePath -> String -> m ()
  readFile :: FilePath -> m (Maybe String)
```